OKTA-631075 Add CLIENT_AS_PRINCIPAL instructions for OAuth Org2Org configuration (#4293)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/secure-oauth-between-orgs/main/index.md

 }' "https://${yourHubOktaDomain}/oauth2/v1/clients"
```

> **Note:** You can copy the entire [response from the previous `GET /api/v1/apps/${id}/credentials/keys` request](#response-example) in to the `jwks.keys` array. The `created`, `lastUpdated`, and `expiresAt` properties are ignored in the POST request. Alternatively, you can remove them from the `jwks.keys` parameter body.

### Assign admin roles to the OAuth 2.0 service app

> > **Note:** These instructions are for Production orgs with the **Assign admin roles to public client apps** feature enabled and Preview orgs.<br><br>

Reviewer: ```suggestion
> **Note:** These instructions are for Preview orgs and Production orgs with the **Assign admin roles to public client apps** feature enabled.<br><br>
```

For the hub-and-spoke OAuth 2.0 Org2Org provisioning connection, Okta recommends that you assign the following [standard admin roles](/docs/concepts/role-assignment/#standard-role-types):

* `USER_ADMIN` (Group Administrator)
* `GROUP_MEMBERSHIP_ADMIN` (Group Membership Administrator)

<!-- Consider adding this section when custom roles can be verified by the Integration Platform team at GA

If you're using [custom roles](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-creating-custom-admin-roles) for the OAuth 2.0 Org2Org provisioning connection, you require the following:

* Permissions required:
  * Manage users
  * Manage groups

* Resource sets recommended:
  * All Users
  * All Groups

> > **Note:** You can bind resource sets to many different types of resources, such as user groups, flows, authorization servers, and so on. You don't necessarily have use the All Groups resource set in this use case. You can specify permissions for a specific subset of users in a custom role. See [Custom role assignment](/docs/concepts/role-assignment/#custom-role-assignment) and the [Custom Role assignment API operations](/docs/reference/api/roles/#custom-role-assignment-operations).

Reviewer: ```suggestion
> **Note:** You can bind resource sets to many different types of resources, such as user groups, flows, authorization servers, and so on. You don't necessarily have to use the All Groups resource set in this use case. You can specify permissions for a specific subset of users in a custom role. See [Custom role assignment](/docs/concepts/role-assignment/#custom-role-assignment) and the [Custom Role assignment API operations](/docs/reference/api/roles/#custom-role-assignment-operations).
```

-->

As an Okta admin, make a `POST /oauth2/v1/clients/${yourServiceAppId}/roles` request to the hub org with the following required parameters to assign an admin role:

| Parameter |  Description/Value   |
| --------- |  ------------- |
| `yourServiceAppId`  |  Specify the `client_id` value from the previous response when the service app was created. In the following role assignment example, the `${yourServiceAppId}` variable name is used instead of `client_id`.|
| `type`  |  Specify the admin role to assign to the service app. Use the recommended standard admin roles (`USER_ADMIN`, `GROUP_MEMBERSHIP_ADMIN`). |

See [Assign a Role to a client application](/docs/reference/api/roles/#assign-a-role-to-a-client-application) in the Role Assignment API reference.

##### Request example

 ```bash
 curl -X POST \
 -H "Accept: application/json" \
 -H "Content-Type: application/json" \
 -H "Authorization: SSWS ${hubApiToken}" \
 -d '{
    "type": "USER_ADMIN"
  }' "https://${yourHubOrgDomain}/oauth2/v1/clients/${yourServiceAppId}/roles"
 ```

> > **Note:** The admin roles determine which resources the actions can be performed on (such as a specific group of users or a specific set of apps), while scopes determine the action that can be performed (such as manage users, read apps, and so on). Therefore, the admin roles need to have enough permissions for the scopes provided.

Reviewer: hoping I didn't mess up the meaning
```suggestion
> **Note:** The admin roles determine which resources the admin can perform the actions on (such as a specific group of users or a specific set of apps), while scopes determine the action that the admin can perform (such as manage users, read apps, and so on). Therefore, the admin roles need to have enough permissions for the scopes provided.
```

> **Note**: You can copy the entire [response from the previous `GET /api/v1/apps/${id}/credentials/keys` request](#response-example) in to the `jwks.keys` array. The `created`, `lastUpdated`, and `expiresAt` properties are ignored in the POST request. Alternatively, you can remove them from the `jwks.keys` parameter body.

#### Assign admin roles to the OAuth 2.0 service app

> > **Note:** Before the **Assign admin roles to public client apps** feature, the Super Administrator (`SUPER_ADMIN`) role was automatically assigned to service apps when they're created. You can now fine-tune the resources that a service app can access by assigning specific standard or custom admin roles. Ensure that the **Assign admin roles to public client apps** Self-Service feature is enabled in your org for granular role assignment. See [Manage Early Access and Beta features](https://help.okta.com/okta_help.htm?id=ext_Manage_Early_Access_features) for more information on the Okta Self-Service Feature Manager.

Reviewer 2: In this note, can we callout that with **Assign admin roles to public client apps**, the role is no longer assigned and needs to be assigned before proceeding further?

Author: ok, i'll add this point.

Assign admin roles for every OAuth 2.0 service app that you create in the hub org. Service apps with assigned admin roles are constrained to the permissions and resources that are included in the role. This improves security for an org since it ensures that service apps only have access to the resources that are needed to perform their tasks. You can assign the [standard admin roles](https://help.okta.com/okta_help.htm?type=oie&id=ext-use-standard-roles) or a [custom admin role](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-creating-custom-admin-roles) with permissions to specific resource sets.

> For the hub-and-spoke OAuth 2.0 Org2Org provisioning connection, Okta recommends that you assign the following [standard admin roles](https://help.okta.com/okta_help.htm?type=oie&id=ext-use-standard-roles):

Reviewer 2: This link is not working for me. Should this link be used instead? https://help.okta.com/en-us/Content/Topics/Security/administrators-admin-comparison.htm

Author: The link should work now. Barbara just added the alias yesterday.
I'll request for another alias for the comparison page and use that link.

> * "USER_ADMIN"

Reviewer 2: These roles should be reviewed the owning team for Org2Org.
I believe USER_ADMIN and GROUP_ADMINISTRATOR are referring to the same role. One is an internal name and the other is an external one. Is it meant to be GROUP_MEMBERSHIP_ADMIN?

Author: Does GROUP_MEMBERSHIP_ADMIN have permission to create and manage users?
The Org2Org service app requires `okta.users.manage` and `okta.groups.manage` scopes, so what standard role can manage both users and admin?

* "GROUP_ADMINISTRATOR"

If you're using [custom roles](https://help.okta.com/okta_help.htm?type=oie&id=ext-about-creating-custom-admin-roles), you require the following for the OAuth 2.0 Org2Org provisioning connection:

> * Permissions required:

Reviewer 2: Same comment about Org2Org team verifying these roles and resource sets.

  * Manage users
  * Manage groups

* Resource sets recommended:
  * All Users
  * All Groups

> > **Note:** You can bind resource sets with specific user groups to a custom role. See [Custom Role assignment operations](/docs/reference/api/roles/#custom-role-assignment-operations).

Reviewer 2: Resource sets can be bound to many different types of resources (user groups, flows, auth servers, etc.). Should we just not mention user groups here?

Author: ok, will do. thanks!

As an Okta admin, make a `POST /oauth2/v1/clients/${yourServiceAppId}/roles` request to the hub org with the following required parameters to assign an admin role:

| Parameter |  Description/Value   |
| --------- |  ------------- |
| `yourServiceAppId`  |  Specify the `client_id` value from the previous response when the service app was created. In the following example, the `${yourServiceAppId}` variable name is used instead of `client_id`.|
| `type`  |  Specify the admin role to assign to the service app. Use the recommended standard admin roles (`USER_ADMIN`, `GROUP_ADMINISTRATOR`) or a custom admin role (`CUSTOM`). |

> See [Assign a Role to a client application](/docs/reference/api/roles/#assign-a-role-to-a-client-application) in the Role Assignment API reference.

Reviewer 3: You could link to the redocly docs here: https://developer.okta.com/docs/api/openapi/okta-oauth/oauth/tag/RoleAssignment/#tag/RoleAssignment/operation/assignRole

Author: Currently, that endpoint doesn't display the Role object in the response, so I'll use the VuePress link until Redocly is fixed.

##### Request example

 ```bash
 curl -X POST \
 -H "Accept: application/json" \
 -H "Content-Type: application/json" \
 -H "Authorization: SSWS ${hubApiToken}" \
 -d '{
    "type": "USER_ADMIN"
  }' "https://${yourHubOrgDomain}/oauth2/v1/clients/${yourServiceAppId}/roles" \
 ```

> > **Note:** The admin roles provide permission to resources, while the scopes provide operations that can be performed on the permissible resources. Therefore, the admin roles need to have enough permissions for the scopes provided. For example: scopes determine the action that can be performed (manage users, read apps, and so on), while admin roles determine which resources the actions can be performed on (a specific group of users).

Reviewer 3: I like how this is written. But my mind expected a certain construction i the last sentence. Something like:

```suggestion
> **Note:** The admin roles provide permission to resources, while the scopes provide operations that can be performed on the permissible resources. Therefore, the admin roles need to have enough permissions for the scopes provided. For example: admin roles determine which resources the actions can be performed on (a specific group of users), while scopes determine the action that can be performed (manage users, read apps, and so on).
```

Author: I like your suggestion. I think the ordering works better :-)
But re-reading it again, I'm thinking of adding the example in the parenthesis (since that's where the examples are).
